js/runner.go

```diff
@@ -312,6 +307,77 @@ func (r *Runner) IsExecutable(name string) bool {
 	return exists
 }
 
+// HandleSummary calls the specified summary callback, if supplied.
+func (r *Runner) HandleSummary(ctx context.Context, summary *lib.Summary) (map[string]io.Reader, error) {
+	summaryDataForJS := summarizeMetricsToObject(summary, r.Bundle.Options)
+
+	out := make(chan stats.SampleContainer, 100)
+	defer close(out)
+
+	go func() { // discard all metrics
+		for range out {
+		}
+	}()
+
+	vu, err := r.newVU(0, out)
+	if err != nil {
+		return nil, err
+	}
+
+	handleSummaryFn := goja.Undefined()
+	if exported := vu.Runtime.Get("exports").ToObject(vu.Runtime); exported != nil {
+		fn := exported.Get(consts.HandleSummaryFn)
+		if _, ok := goja.AssertFunction(fn); ok {
+			handleSummaryFn = fn
+		} else if fn != nil && !goja.IsUndefined(fn) && !goja.IsNull(fn) {
+			return nil, fmt.Errorf("exported identfier %s must be a function", consts.HandleSummaryFn)
+		}
+	}
+	ctx = common.WithRuntime(ctx, vu.Runtime)
+	ctx = lib.WithState(ctx, vu.state)
+	ctx, cancel := context.WithTimeout(ctx, r.getTimeoutFor(consts.HandleSummaryFn))
+	defer cancel()
+	go func() {
+		<-ctx.Done()
+		vu.Runtime.Interrupt(context.Canceled)
+	}()
+	*vu.Context = ctx
+
+	handleSummaryWrapperRaw, err := vu.Runtime.RunString(summaryWrapperLambdaCode)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error while getting the summary wrapper: %w", err)
+	}
+	handleSummaryWrapper, ok := goja.AssertFunction(handleSummaryWrapperRaw)
+	if !ok {
+		return nil, fmt.Errorf("unexpected error did not get a callable summary wrapper")
+	}
+
+	wrapperArgs := []goja.Value{
+		handleSummaryFn,
+		vu.Runtime.ToValue(r.Bundle.RuntimeOptions.SummaryExport.String),
+		vu.Runtime.ToValue(summaryDataForJS),
+		vu.Runtime.ToValue(getOldTextSummaryFunc(summary, r.Bundle.Options)), // TODO: remove
+	}
+	rawResult, _, _, err := vu.runFn(ctx, false, handleSummaryWrapper, wrapperArgs...)
+
+	// TODO: refactor the whole JS runner to avoid copy-pasting these complicated bits...
```

Comment:
I am pretty sure you can just use `runPart` for most of the above. 
You will need to keep the check for whether `exports.handleSummary` is a function, but everything else seems to be done by `runPart` either way ... I think ...

Reply:
Although I guess you will need to move https://github.com/loadimpact/k6/blob/0838813db6edead0f6faa3cdaadf3b5eb070a1ab/js/summary.go#L95-L118 to golang code after the `runPart`

Reply:
Yeah, I initially tried using `runPart`, but it was an even bigger mess. We can figure out some way to refactor the whole `js.Runner` internally in the future to reduce the boilerplate, but for now I don't want to mess with it more if I don't need to... :sweat_smile: